frame/election-provider-multi-phase/src/signed.rs

```diff
@@ -0,0 +1,828 @@
+	}
+
+	/// Remove the signed submission with the highest score from the set.
+	pub fn pop_last(&mut self) -> Option<SignedSubmissionOf<T>> {
+		let (highest_score, idx) = self.indices.iter().rev().next()?;
+		let (highest_score, idx) = (*highest_score, *idx);
+		self.indices.remove(&highest_score);
+		self.take_submission(idx)
+	}
+}
+
+impl<T: Config> Deref for SignedSubmissions<T> {
+	type Target = SubmissionIndicesOf<T>;
+
+	fn deref(&self) -> &Self::Target {
+		&self.indices
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	/// `Self` accessor for `SignedSubmission<T>`.
+	pub fn signed_submissions() -> SignedSubmissions<T> {
+		SignedSubmissions::<T>::get()
+	}
+
+	/// Finish the signed phase. Process the signed submissions from best to worse until a valid one
+	/// is found, rewarding the best one and slashing the invalid ones along the way.
+	///
+	/// Returns true if we have a good solution in the signed phase.
+	///
+	/// This drains the [`SignedSubmissions`], potentially storing the best valid one in
+	/// [`QueuedSolution`].
+	pub fn finalize_signed_phase() -> (bool, Weight) {
+		let mut all_submissions = Self::signed_submissions();
+		let mut found_solution = false;
+		let mut weight = T::DbWeight::get().reads(1);
+
+		while let Some(best) = all_submissions.pop_last() {
+			let SignedSubmission { solution, who, deposit, reward } = best;
+			let active_voters = solution.compact.voter_count() as u32;
+			let feasibility_weight = {
+				// defensive only: at the end of signed phase, snapshot will exits.
+				let SolutionOrSnapshotSize { voters, targets } =
+					Self::snapshot_metadata().unwrap_or_default();
+				let desired_targets = Self::desired_targets().unwrap_or_default();
+				T::WeightInfo::feasibility_check(
+					voters,
+					targets,
+					active_voters,
+					desired_targets,
+				)
+			};
+			// the feasibility check itself has some weight
+			weight = weight.saturating_add(feasibility_weight);
+			match Self::feasibility_check(solution, ElectionCompute::Signed) {
+				Ok(ready_solution) => {
+					Self::finalize_signed_phase_accept_solution(
+						ready_solution,
+						&who,
+						deposit,
+						reward,
+					);
+					found_solution = true;
+
+					weight = weight
+						.saturating_add(T::WeightInfo::finalize_signed_phase_accept_solution());
+					break;
+				}
+				Err(_) => {
+					Self::finalize_signed_phase_reject_solution(&who, deposit);
+					weight = weight
+						.saturating_add(T::WeightInfo::finalize_signed_phase_reject_solution());
+				}
+			}
+		}
+
+		// Any unprocessed solution is pointless to even consider. Feasible or malicious,
+		// they didn't end up being used. Unreserve the bonds.
+		let discarded = all_submissions.len();
+		for SignedSubmission { who, deposit, .. } in all_submissions.drain() {
+			let _remaining = T::Currency::unreserve(&who, deposit);
+			weight = weight.saturating_add(T::DbWeight::get().writes(1));
+			debug_assert!(_remaining.is_zero());
+		}
+
+		all_submissions.put();
```

Comment:
at this point all internal storage items of `all_submissions` must have been cleared. I recommend to either explicitly clear them, or preferably combine with debug-assert both:
```suggestion
		all_submissions.put();
		debug_assert!(!SignedSubmissionIndices::<T>::exists());
		debug_assert!(!SignedSubmissionNextIndex::<T>::exists());
		dbueg_assert!(!SignedSubmissionsMap::<T>::exists());
```

Reply:
That's a really good idea: 2c61152ab.